src/ddspdrum/module.py

```diff
@@ -0,0 +1,237 @@
+"""
+Synth modules.
+
+    TODO    :   - VCA is slow. Blocks? Not sure what's best for tensors.
+                - Convert operations to tensors, obvs.
+"""
+
+import numpy as np
+from scipy.signal import resample
+
+from ddspdrum.defaults import CONTROL_RATE, SAMPLE_RATE
+
+
+class SynthModule:
+    """
+    Base class for synthesis modules. Mostly helper functions for the moment.
+    """
+
+    def __init__(
+        self, sample_rate: int = SAMPLE_RATE, control_rate: int = CONTROL_RATE
+    ):
+        self.sample_rate = SAMPLE_RATE
+        self.control_rate = CONTROL_RATE
+
+    def control_to_sample_rate(self, control: np.array) -> np.array:
+        """
+        Resample a control signal to the sample rate
+        TODO: One thing I worry about with all these casts to
+        ints back and forth between sample and control rate is
+        off-by-one errors.
+        I'm beginning to believe we should standardize as much
+        as possible on nsamples in most places instead of
+        float duration in seconds, and just convert to ncontrol
+        when needed..
+        """
+        # Right now it appears that all signals are 1d, but later
+        # we'll probably convert things to 2d: instance x signal
```

Comment:
So this is my current assumption about how I as a user would use this library, but tell me if you have other patterns in mind that we should have in the back of our minds in the roadmap:

* You define a particular synth architecture i.e. module layout (or maybe have several you are comparing side by side).
* You want to do some kind of training or inference (whether its synthesis or inverse-synthesis) in *mini-batches* not one patch at a time.
* If we standardize upon this idea that everything will be mini-batches, it's a bit simpler than the numpy style code-base where they constantly check if everything is one or two dimensions. The only real pain for the user is if they do want send one single patch through, in which they have to create a single row matrix.
* It's a little fiddly to convert from 1d to 2d and get everything right, so if we agree I think we fix this in the near future.
* Each module with then not have one set of parameters, but `batch_size` sets of parameters. This is a bit unusual for neural networks, that the network parameters depend upon the `batch_size`, but it's a pattern that seems to make the most sense to me given the few DDSP and inverse synth nets I've tried to build.

Reply:
Interesting! I was thinking about this a bit more as a traditional synth where you define your architecture (we can have pre-defined ones such as that Axxe drum machine), then you instantiate that and can set / update parameters on that one object and play sounds on that. 

But I think this is a bit different than what you are suggesting? So instead of having multiple instances of a synths within a batch, we would have one instance that can handle batch process -- And each module would need to support having mini-batches of parameters as well? We may have to decide on a simple parameter data structure to help keep this organized. If we go this route I would want to make sure that it is easy to use with a single parameter setting because that will be more intuitive for users coming from a music / synth background. I'm all for it if it will make the neural net stuff easier!

Reply:
My feeling, let's make sure it works in 1D before giving ourselves the headache. If it works slow we can then make it fast.

Reply:
Okay, let's leave everything as 1D for now.

@jorshi I agree that multiple architectures is very important. One of the potentially cool things about what we're doing is architecture search. Traditional drum synths don't sound like real synths. I'm wondering how far we can push this towards natural drums by having complicated architecture. I'm not talking full on inception architecture stuff, but definitely more complex than usual. What's interesting about the DDSP approach is that in principle, if we can train neural networks on it, we can use the neural networks to program very complicated architectures to get the sounds we want.

Reply:
@turian Multiple subtractive synth architectures does not a real sound make. If you add up four motorcycles, do you get a car?

Reply:
@maxsolomonhenry not if you *add* them but if you chop them up and re-assemble them then maybe you do get a car. I'm not proposing linear regression on motorcycles. I'm proposing passing them through several non-linear operations.